awx_collection/plugins/modules/tower_inventory_source.py

```diff
@@ -20,333 +20,347 @@
+                                                'update_on_project_launch',
+                                                'source_regions',
+                                                'instance_filters',
+                                                'group_by',
+                                                'source_vars',
+                                                'source_script'
+                                                ]),
+                             ('source', 'tower', ['source_project',
+                                                  'source_path',
+                                                  'update_on_project_launch',
+                                                  'source_regions',
+                                                  'group_by',
+                                                  'source_vars',
+                                                  'source_script'
+                                                  ]),
+                             ('source', 'custom', ['source_project',
+                                                   'source_path',
+                                                   'update_on_project_launch',
+                                                   'source_regions',
+                                                   'instance_filters',
+                                                   'group_by'
+                                                   ]),
+                         ])
+
+    optional_vars = {}
+    # Extract our parameters
     name = module.params.get('name')
+    new_name = module.params.get('new_name')
+    optional_vars['description'] = module.params.get('description')
     inventory = module.params.get('inventory')
-    source = module.params.get('source')
+    optional_vars['source'] = module.params.get('source')
+    optional_vars['source_path'] = module.params.get('source_path')
+    source_script = module.params.get('source_script')
+    optional_vars['source_vars'] = module.params.get('source_vars')
+    credential = module.params.get('credential')
+    optional_vars['source_regions'] = module.params.get('source_regions')
+    optional_vars['instance_filters'] = module.params.get('instance_filters')
+    optional_vars['group_by'] = module.params.get('group_by')
+    optional_vars['overwrite'] = module.params.get('overwrite')
+    optional_vars['overwrite_vars'] = module.params.get('overwrite_vars')
+    optional_vars['custom_virtualenv'] = module.params.get('custom_virtualenv')
+    optional_vars['timeout'] = module.params.get('timeout')
+    optional_vars['verbosity'] = module.params.get('verbosity')
+    optional_vars['update_on_launch'] = module.params.get('update_on_launch')
+    optional_vars['update_cache_timeout'] = module.params.get('update_cache_timeout')
+    source_project = module.params.get('source_project')
+    optional_vars['update_on_project_update'] = module.params.get('update_on_project_update')
```

Comment:
On the principle of consistency, I'm not liking this `option_vars` dict. If I read it right, the difference between it and the `*_fields` dict is that it holds falsy values. That's not something that we need, I'd rather just test falsyness before we put it into the `inventory_source_fields`

Again, the thing I really want to double-down on here is that the modules should all look very consistent, and this is one I really want to bring into line. Otherwise we lose out on the benefits from a lot of the rest of the work.

Reply:
For this one, this was kind of a "hack" because of the issues with the API in regards to inventory sources. The only thing we know we will have to pass back to the API is in the lines:
```
    inventory_source_fields = {
        'name': new_name if new_name else name,
        'inventory': inventory_id,
    }
```

All of the other parameters may or may not be present based on the inventory source type. So, instead of doing a bunch of
```
if <param> is not None:
    inventory_source_fields[<param>] = <param_varaibale>
```

I threw everything into the `optional_vars` dict and then down below:
```
    # Layer in all remaining optional information
    for field_name in optional_vars:
        if optional_vars[field_name]:
            inventory_source_fields[field_name] = optional_vars[field_name]
```

We have on the list to file a bug against the API but until that is resolved this just saves us a bunch of lines of code.

With that said, I see your point about the patten we could add the additional lines instead of the dict /w for.

Reply:
What bug would you file against the API? The canonical behavior with regards to Django and DRF is that CharField type will not accept null values, you have to use a null-char-field, which is a very recent construct, and we're not using that for these fields. Several of the fields should be blank under normal conditions. But handling `None` is something that all modules should be obligated to do - because tower-cli did this. It popped the `None` values out of the dict before it used it in data to the API. It is true that some fields are nullable, but we honestly could never really well handle the act of nulling it (I don't expect that we will start).

So if non-provided values are actually passed to the API as null, that could be an issue for _all the other modules_ because when you mention "doing a bunch of" checking if the params are null, that's something I expect to see. Now we can clean the code up, just loop over a tuple of the field names or something, so an extra check per-field doesn't requires a ton more code.

I hope I'm making sense.